Fail restore when the shard allocations max retries count is reached

tlrx opened the pull request:
When the allocation of a shard has been retried too many times, the
`MaxRetryDecider` is engaged to prevent any future allocation of the
failed shard. If it happens while restoring a snapshot, the restore
hangs and never completes because it stays around waiting for the
shards to be assigned. It also blocks future attempts to restore the
snapshot again.

This commit changes the current behaviour in order to fail the restore if
a shard reached the maximum allocations attempts without being successfully
assigned.

This is the second part of the #26865 issue.

ywelsch commented on core/src/main/java/org/elasticsearch/snapshots/RestoreService.java:
```diff
@@ -579,23 +573,54 @@ private Updates changes(Snapshot snapshot) {
+                                // check if the maximum number of attempts to restore the shard has been reached. If so, we can fail
+                                // the restore and leave the shards unassigned.
+                                IndexMetaData indexMetaData = metaData.getIndexSafe(unassignedShard.getKey().getIndex());
+                                int maxRetry = MaxRetryAllocationDecider.SETTING_ALLOCATION_MAX_RETRY.get(indexMetaData.getSettings());
```
I think the solution needs to be more generic than depending on the settings of specific allocation deciders. I think we can use `unassignedInfo.getLastAllocationStatus` for that and check if it is `DECIDERS_NO`.

tlrx replied:
That's a good suggestion as I suppose that a restore can also be stuck because the deciders cannot assign the shard (no enough space on disk, awareness rules forbid allocation etc). I also like it to be more generic.

I think I can give it a try by reverted portion of code and override `unassignedInfoUpdated()`... I'll push something if it works.

tlrx commented:
Thanks for your reviews. I think @ywelsch is right, we can be more generic by checking the last allocation status and fail the restore if needed. 

I updated the code to revert some changes I did and use the last allocation status instead. I also added another test based on shard allocation filtering settings that prevent the shards to be assigned. Please let me know what you think :)

ywelsch commented on core/src/main/java/org/elasticsearch/snapshots/RestoreService.java:
```diff
@@ -569,6 +568,21 @@ public void shardInitialized(ShardRouting unassignedShard, ShardRouting initiali
+                    if (newUnassignedInfo.getLastAllocationStatus() == UnassignedInfo.AllocationStatus.DECIDERS_NO) {
+                        Snapshot snapshot = ((SnapshotRecoverySource) recoverySource).snapshot();
+                        String reason = "shard was denied allocation by all allocation deciders";
+                        changes(snapshot).unassignedShards.put(unassignedShard.shardId(),
```
I think instead of adding more types here (unassignedShards), better we do the reverse and fold failedShards, startedShards and unassignedShards into just "updates".
It's not worth separating them just to have this one assertion I've put there.

tlrx replied:
Pff I should have seen that... I agree, that would be better, thanks :)

ywelsch commented on core/src/main/java/org/elasticsearch/snapshots/RestoreService.java:
```diff
@@ -569,6 +568,21 @@ public void shardInitialized(ShardRouting unassignedShard, ShardRouting initiali
+                if (recoverySource.getType() == RecoverySource.Type.SNAPSHOT) {
+                    if (newUnassignedInfo.getLastAllocationStatus() == UnassignedInfo.AllocationStatus.DECIDERS_NO) {
+                        Snapshot snapshot = ((SnapshotRecoverySource) recoverySource).snapshot();
+                        String reason = "shard was denied allocation by all allocation deciders";
```
I would just put "shard could not be allocated on any of the nodes"

ywelsch commented on core/src/test/java/org/elasticsearch/snapshots/SharedClusterSnapshotRestoreIT.java:
```diff
@@ -899,6 +903,232 @@ public void testDataFileCorruptionDuringRestore() throws Exception {
+                                              .setSettings(Settings.builder().put("location", repositoryLocation)));
+
+        // create a test index
+        final String indexName = randomAlphaOfLength(10).toLowerCase(Locale.ROOT);
```
just choose a fixed index name, no need for randomization here :)

ywelsch commented on core/src/test/java/org/elasticsearch/snapshots/SharedClusterSnapshotRestoreIT.java:
```diff
@@ -899,6 +903,232 @@ public void testDataFileCorruptionDuringRestore() throws Exception {
+
+        // attempt to restore the snapshot
+        RestoreSnapshotResponse restoreResponse = client().admin().cluster().prepareRestoreSnapshot("test-repo", "test-snap")
+                                                                            .setMasterNodeTimeout("30s")
```
isn't this the default?

tlrx replied:
It is, thanks

ywelsch commented on core/src/test/java/org/elasticsearch/snapshots/SharedClusterSnapshotRestoreIT.java:
```diff
@@ -899,6 +903,232 @@ public void testDataFileCorruptionDuringRestore() throws Exception {
+        assertBusy(() -> {
+            Client client = client();
+            for (int shardId = 0; shardId < numShards.numPrimaries; shardId++) {
+                ClusterAllocationExplainResponse allocationResponse = client.admin().cluster().prepareAllocationExplain()
```
no need for allocation explain API. you can check all this directly on the cluster state that you get below. I also think that assertBusy won't be needed then.

tlrx replied:
Right, we can use the cluster state in this case and it's even easier. But it seems that the assertBusy() is still needed as the cluster state change can take few miliseconds to propagate.

ywelsch commented on core/src/test/java/org/elasticsearch/snapshots/SharedClusterSnapshotRestoreIT.java:
```diff
@@ -899,6 +903,232 @@ public void testDataFileCorruptionDuringRestore() throws Exception {
+            // Option 2: delete the index and restore again
+            assertAcked(client().admin().indices().prepareDelete(indexName));
+            restoreResponse = client().admin().cluster().prepareRestoreSnapshot("test-repo", "test-snap")
+                                                        .setMasterNodeTimeout("30s")
```
idem

ywelsch commented on core/src/test/java/org/elasticsearch/snapshots/SharedClusterSnapshotRestoreIT.java:
```diff
@@ -899,6 +903,232 @@ public void testDataFileCorruptionDuringRestore() throws Exception {
+        }
+
+        // Wait for the shards to be assigned
+        waitForRelocation();
```
just ensureGreen()

ywelsch commented on core/src/test/java/org/elasticsearch/snapshots/SharedClusterSnapshotRestoreIT.java:
```diff
@@ -899,6 +903,232 @@ public void testDataFileCorruptionDuringRestore() throws Exception {
+     * Test that restoring an index with shard allocation filtering settings that prevents
+     * its allocation does not hang indefinitely.
+     */
+    public void testUnrestorableIndexDuringRestore() throws Exception {
```
I think it's possible to share most of the code with the previous test by calling a generic method with two parameters:
- restoreIndexSettings (which would set maxRetries in first case and filters in second case)
- fixupAction (the action to run to fix the issue)

tlrx replied:
I tried to do something like that, please let me know what you think.

tlrx commented:
Thanks for the reviews! I updated the code, can you please have another look?

tlrx commented:
@ywelsch I added the allocation decider we talked about and I had to rebase and adapt an existing test. The CI is still failing but I'm not sure it is related to my change, I'm still digging. Anyway, I'd be happy if you can have a look, thanks!

ywelsch commented on core/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/RestoreInProgressAllocationDecider.java:
```diff
@@ -0,0 +1,83 @@
+                if (restoreInProgress.snapshot().equals(snapshot)) {
+                    assert restoreInProgress.state().completed() == false : "completed restore should have been removed from cluster state";
+                    RestoreInProgress.ShardRestoreStatus shardRestoreStatus = restoreInProgress.shards().get(shardRouting.shardId());
+                    if (shardRestoreStatus.state() != RestoreInProgress.State.FAILURE) {
```
I would write this check as
```
if (shardRestoreStatus.state().completed() == false) {
```

and then add an assertion that `shardRestoreStatus.state() != SUCCESS` (as the shard should have been moved to started and the recovery source cleaned up at that point).

tlrx replied:
> just wondering if it's possible for shardRestoreStatus to be null.
> I think it can be if you restore from a snapshot, then the restore fails, and you retry another restore with a different subset of indices from that same snapshot.

Good catch, thanks!

ywelsch commented on core/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/RestoreInProgressAllocationDecider.java:
```diff
@@ -0,0 +1,83 @@
+    }
+
+    @Override
+    public Decision canAllocate(final ShardRouting shardRouting, final RoutingNode node, final RoutingAllocation allocation) {
```
can you also  add

```
@Override
    public Decision canForceAllocatePrimary(ShardRouting shardRouting, RoutingNode node, RoutingAllocation allocation) {
        assert shardRouting.primary() : "must not call canForceAllocatePrimary on a non-primary shard " + shardRouting;
        return canAllocate(shardRouting, node, allocation);
    }
```

as this is a hard constraint with no exceptions

tlrx replied:
Good catch

ywelsch commented on core/src/main/java/org/elasticsearch/snapshots/RestoreService.java:
```diff
@@ -563,11 +562,26 @@ public void shardInitialized(ShardRouting unassignedShard, ShardRouting initiali
 
+        @Override
+        public void unassignedInfoUpdated(ShardRouting unassignedShard, UnassignedInfo newUnassignedInfo) {
+            if (unassignedShard.primary()) {
```
only primaries can have a snapshot recovery source, so no need for this extra check here.

tlrx commented on core/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/RestoreInProgressAllocationDecider.java:
```diff
@@ -0,0 +1,83 @@
+
+        final RecoverySource recoverySource = shardRouting.recoverySource();
+        if (recoverySource == null || recoverySource.getType() != RecoverySource.Type.SNAPSHOT) {
+            return allocation.decision(Decision.YES, NAME, "shard is not being restored");
```
Right, thanks

tlrx commented on core/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/RestoreInProgressAllocationDecider.java:
```diff
@@ -0,0 +1,83 @@
+        if (restoresInProgress != null) {
+            for (RestoreInProgress.Entry restoreInProgress : restoresInProgress.entries()) {
+                if (restoreInProgress.snapshot().equals(snapshot)) {
+                    assert restoreInProgress.state().completed() == false : "completed restore should have been removed from cluster state";
```
The assertion asserts that the restore in progress for the current allocation is **not** completed, so I think it's good? It will be marked later as you noticed.

tlrx replied:
Note: we talked about this and Yannick is right, this assertion can be problematic on more busy clusters if a reroute kicks in between the moment the restore completes and the moment the restore is removed from the cluster state by the CleanRestoreStateTaskExecutor

tlrx commented:
@ywelsch Thanks for the review! I updated the code.

ywelsch commented on core/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/RestoreInProgressAllocationDecider.java:
```diff
@@ -0,0 +1,85 @@
+                if (restoreInProgress.snapshot().equals(snapshot)) {
+                    RestoreInProgress.ShardRestoreStatus shardRestoreStatus = restoreInProgress.shards().get(shardRouting.shardId());
+                    if (shardRestoreStatus != null && shardRestoreStatus.state().completed() == false) {
+                        assert shardRestoreStatus.state() != RestoreInProgress.State.SUCCESS : "expected initializing shard";
```
can you also add the shardRestoreStatus state and the shard routing to the failure message here?

tlrx replied:
Sure, will do once the current CI build is finished

tlrx commented:
Thanks @ywelsch! 

@imotov This pull request changed since you LGTMd it, would you like to have another look?

tlrx commented:
Thanks a lot @ywelsch and @imotov. If you're ok I'd like to keep this in master for few days before backporting. I'd like to see it run multiple times on our CI.

tlrx commented:
Backported to 6.x,6.1,6.0 and 5.6
